### PR TITLE
[added] url parameter rules

### LINKED ISSUE #2286: Customizable URL matching rules
Lifting this from #2284:

-----

@mjackson 

Sure - what I mean is that Flask/Werkzeug offer an example of a good abstraction of pluggable pattern matching (via a mini-DSL) that has worked pretty well for me. The way it looks like is that instead of e.g. our

```
/a/:foo/b
```

The preferred syntax for specifying a URL rule looks like e.g.

```
/a/<foo>/b
```

This is extensible to allow people to do things like

```
/a/<str(length=2):foo>/b
/a/<int:foo>/b
/a/<path:foo>/b
/a/<uuid:foo>/b
/a/<my_custom_rule:foo>/b
```

`str`, `int` and `path` are built-in (in this case the URL rules handle both matching and converting params), while `my_custom_rule` is an example of the syntax for using a custom URL matching rule. Per the `str(length=2)` example these converters can also take arguments.

These rules are used to both match and generate URLs.

The nice thing about something like this sort of general solution (not necessarily in this exact syntax form) is that they'd let users solve their own problems, while still keeping the same workflow around route matching, instead of adding something like `next()` from Express.

ETA: Werkzeug docs are here: http://werkzeug.pocoo.org/docs/0.10/routing/

-----

The `<int:foo>` case is one decent motivation for having something like this BTW - I've found it a nice convenience to have my URL routing framework just 404 anything that doesn't match the pattern I want before it even gets to my URL handler, and if I'm specifying that it's an integer in the URL rule, I may as well also have the routing convert that parameter to the right type (rather than just a string).

Also I forgot to say, but the difference between `str` and `path` above is that `path` is like the `**` glob in that it also matches `/`s as needed.

### PR DESCRIPTION
**Pull request by itajaja**
Closes #2286 

Another pull request for custom matching parameter rules, this time with the specifics proposed by @th0r

This time was easier, and I like the final result more. I still need to add the documentation and the rule inheritance feature, but let men know what you think for now

### CODE REVIEW

**Comment by itajaja** [comment]
I added the nesting as well. I don't know how much is useful though, as @taion said, you want to put the rule on the same component it's declared, for readability, but it wasn't hard to implement so I put it. Let me know what you think, I am working on updating the doc in the meantime.

**Comment by taion** [comment]
Haven't had a chance to look at this yet, but should we close #2402?

**Comment by itajaja** [comment]
@taion If we all agree on this syntax, I would say so :)

**Comment by taion** [comment]
I think in a vacuum I'd have preferred the Flask-style syntax for specifying URL variables (wrapped in angle brackets rather than prefixed with a colon - but probably still with the external type specification), but given what we already have, this is better.

File: modules/RouteUtils.js [hunk 1]
```diff
@@ -27,10 +39,14 @@ function createRoute(defaultProps, props) {
   return { ...defaultProps, ...props }
 }
 
-export function createRouteFromReactElement(element) {
+export function createRouteFromReactElement(element, parentRoute) {
   const type = element.type
   const route = createRoute(type.defaultProps, element.props)
 
+  if(parentRoute && parentRoute.params && route.params) {
+    route.params = mergeObjects(parentRoute.params, route.params)
```
Comment (taion) [inline modules/RouteUtils.js]: You can just use ES6 object spread stuff to write this as

```js
route.params = {...parentRoute.params, ...route.params}
```

**Comment by taion** [comment]
I haven't had time to look at this in depth, but I really think the right way to do this architecturally per @mjackson's comment is to split out something with an API that looks like `matchPattern` into a separate library.

**Comment by itajaja** [comment]
You mean like a separate repo right? So we aim at refactoring out `RouteUtils`, with caching and all? Before embarking in that, I want to make sure we are in-line. suggestions (naming, scope, etc.) are welcome

**Comment by taion** [comment]
You mean `PatternUtils`, right? That's my thought - pattern-matching is not React Router specific.

**Comment by itajaja** [comment]
whops, `patternUtils`. Let's see what I am able to do.

**Comment by taion** [comment]
I'm closing this PR for now since the development is in a separate repo.

**Pull request closed without merging**